storage/aptosdb/src/db_debugger/truncate/mod.rs

```diff
@@ -0,0 +1,567 @@
+
+        Ok(())
+    }
+
+    fn delete_per_epoch_data(
+        ledger_db: &DB,
+        start_version: Version,
+        end_version: Version,
+        batch: &SchemaBatch,
+    ) -> Result<()> {
+        let mut iter = ledger_db.iter::<LedgerInfoSchema>(ReadOptions::default())?;
+        iter.seek_to_last();
+        if let Some((epoch, ledger_info)) = iter.next().transpose()? {
+            let version = ledger_info.commit_info().version();
+            assert_lt!(version, end_version);
+            if version >= start_version {
+                batch.delete::<LedgerInfoSchema>(&epoch)?;
+            }
+        }
+
+        let mut iter = ledger_db.iter::<EpochByVersionSchema>(ReadOptions::default())?;
+        iter.seek(&start_version)?;
+
+        for item in iter {
+            let (version, epoch) = item?;
+            assert_lt!(version, end_version);
+            batch.delete::<EpochByVersionSchema>(&version)?;
+            batch.delete::<LedgerInfoSchema>(&epoch)?;
+        }
+
+        Ok(())
+    }
+
+    fn delete_per_version_data(
+        start_version: Version,
+        end_version: Version,
+        batch: &SchemaBatch,
+    ) -> Result<()> {
+        for version in start_version..end_version {
+            batch.delete::<TransactionInfoSchema>(&version)?;
+            batch.delete::<TransactionSchema>(&version)?;
+            batch.delete::<VersionDataSchema>(&version)?;
+            batch.delete::<WriteSetSchema>(&version)?;
+        }
+
+        Ok(())
+    }
+
+    fn delete_state_value_and_index(
+        ledger_db: &DB,
+        start_version: Version,
+        end_version: Version,
+        batch: &SchemaBatch,
+    ) -> Result<()> {
+        let mut iter = ledger_db.iter::<StaleStateValueIndexSchema>(ReadOptions::default())?;
+        iter.seek(&start_version)?;
+
+        for item in iter {
+            let (index, _) = item?;
+            assert_lt!(index.stale_since_version, end_version);
+            batch.delete::<StaleStateValueIndexSchema>(&index)?;
+            batch.delete::<StateValueSchema>(&(index.state_key, index.stale_since_version))?;
+        }
+
+        Ok(())
+    }
+
+    fn truncate_state_merkle_db(state_merkle_db: &DB, target_version: Version) -> Result<()> {
+        let status = StatusLine::new(Progress::new(target_version));
+        loop {
+            let batch = SchemaBatch::new();
+            let current_version = Self::get_current_version_in_state_merkle_db(&state_merkle_db)?
+                .expect("Current version of state merkle db must exist.");
+            status.set_current_version(current_version);
+            assert_ge!(current_version, target_version);
+            if current_version == target_version {
+                break;
+            }
+
+            let mut iter =
+                state_merkle_db.iter::<JellyfishMerkleNodeSchema>(ReadOptions::default())?;
+            iter.seek(&NodeKey::new_empty_path(current_version))?;
```

Comment:
is this supposed to be target_version? and deleting from root to current version

Reply:
No. The intention is to delete the current (largest) root together with all nodes generated in this version in a single db batch.

Note we do this in a loop, repeat until the current_version is equal to the target_version. The motivation here is to avoid the db batch to be too large.